### PR TITLE
Add some animation playback mode

### PR DESCRIPTION
**Pull request by Insofan**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description

About [issue #3105 ](https://github.com/SDWebImage/SDWebImage/issues/3105)

### CODE REVIEW

File: SDWebImage/Core/SDAnimatedImageView.h [hunk 1]
```diff
@@ -52,6 +53,10 @@
  `< 0.0` is not supported currently and stop animation. (may support reverse playback in the future)
  */
 @property (nonatomic, assign) double playbackRate;
+
+/// Asynchronous setup animation playback mode. Default mode is SDAnimatedImagePlaybackModeNormal.
+@property (nonatomic, assign) SDAnimatedImagePlaybackMode playbackMode;
+
```
Comment (dreampiggy) [inline SDWebImage/Core/SDAnimatedImageView.h:59]: We can expose the `SDAnimatedImagePlayer` this property into the open API of SDAnimatedImageView. With some code comment such as `Setting directly to the player will cause some animation status not get synchronized. This property is used for some advanced usage or debugging usage.`

File: SDWebImage/Core/SDAnimatedImagePlayer.m [hunk 2]
```diff
@@ -142,6 +143,14 @@ - (void)setupCurrentFrame {
...
     }
+    if (self.playbackMode == SDAnimatedImagePlaybackModeNormal ||
+        self.playbackMode == SDAnimatedImagePlaybackModeBounce) {
+        self.currentFrameIndex = 0;
+    } else if (self.playbackMode == SDAnimatedImagePlaybackModeReverse ||
+               self.playbackMode == SDAnimatedImagePlaybackModeReversedBounce) {
+        self.currentFrameIndex = self.totalFrameCount - 1;
+    }
+    
     if ([self.animatedProvider isKindOfClass:[UIImage class]]) {
```
Comment (dreampiggy) [inline SDWebImage/Core/SDAnimatedImagePlayer.m:154]:
```suggestion
    if (!self.currentFrame) {
        if ([self.animatedProvider isKindOfClass:[UIImage class]]) {
    }
```

Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImagePlayer.m:154]: This is the old logic from `setupCurrentFrame`

File: SDWebImage/Core/SDAnimatedImageView.h (hunk 1 above)
Comment (Insofan) [reply SDWebImage/Core/SDAnimatedImageView.h:59]: If expose the `SDAnimatedImagePlayer ` ,`SDAnimatedImageView`'s some property like` playbackRate` and `maxBufferSize` should be hidden？

Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImageView.h:59]: Just keep both. This is stable API from 5.0...

Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImageView.h:59]: In the future maybe we can remove the same API exposed to the ImageView && Player.

File: SDWebImage/Core/SDAnimatedImageView.h [hunk 3]
```diff
@@ -20,6 +20,10 @@
  For AppKit: use `-setAnimates:` to control animating, `animates` to check animation state. This view is layer-backed.
  */
 @interface SDAnimatedImageView : UIImageView
+/**
+ The animation player.
+ */
+@property (nonatomic, strong, nonnull) SDAnimatedImagePlayer *player;
```
Comment (dreampiggy) [inline SDWebImage/Core/SDAnimatedImageView.h:26]: readonly...

Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImageView.h:26]: And, this is nullable.

```
/**
 The internal animation player.
 This property is only used for advanced usage, like inspecting/debugging animation status, control progressive loading, complicated animation frame index control, etc.
 @warning pay attention if you directly update the player's property like `totalFrameCount`, `totalLoopCount`, the same property on SDAnimatedImageView will not sync back as player.
 */
@property (nonatomic, strong, nonnull) SDAnimatedImagePlayer *player;
```

File: SDWebImage/Core/SDAnimatedImageView.m [hunk 4]
```diff
@@ -239,6 +242,16 @@ - (double)playbackRate
     return _playbackRate;
 }
 
+- (void)setPlaybackMode:(SDAnimatedImagePlaybackMode)playbackMode {
+    _playbackMode = playbackMode;
+    self.player.playbackMode = playbackMode;
+}
+
+- (SDAnimatedImagePlaybackMode)playbackMode {
```
Comment (dreampiggy) [inline SDWebImage/Core/SDAnimatedImageView.m:250]: Write like other property:


```objective-c
- (double)playbackRate
{
    if (!_initFinished) { 
        return SDAnimatedImagePlaybackModeNormal; // Defaults to .nomal
    }
    return _playbackRate;
}
```

**Review by dreampiggy — CHANGES_REQUESTED** [review]

**Review by dreampiggy — APPROVED** [review]

**Comment by dreampiggy** [comment]
This MR can be merged. The test case seems not so solid on tvOS, ignore.

**Pull request merged**